src/mergeDeep.ts

```diff
@@ -0,0 +1,117 @@
+import { purry } from './purry';
+
+type UnknownRecordOrArray =
+  | Record<PropertyKey, unknown>
+  | ReadonlyArray<unknown>;
+
+function isRecordOrArray(object: unknown): object is UnknownRecordOrArray {
+  return (
+    Array.isArray(object) ||
+    (typeof object === 'object' &&
+      object !== null &&
+      Object.getPrototypeOf(object) === Object.prototype)
+  );
+}
+
+/**
+ * Recursively merges two values, `a` and `b`.
+ * @param target - The first value to merge. This is the dominant parameter in the merge operation.
+ * @param source - The second value to merge.
+ * @signature
+ *    R.mergeDeep(a, b)
+ * @example
+ *    R.mergeDeep({ foo: 'bar', x: 1 }, { foo: 'baz', y: 2 }) // => { foo: 'bar', x: 1, y: 2 }
+ * @data_first
+ * @category Object
+ * @pipeable
+ */
+export function mergeDeep<
+  Target extends UnknownRecordOrArray,
+  Source extends UnknownRecordOrArray
+>(
+  target: Target,
+  source: Source
+): MergeDeep<Target, Source, { arrayMergeMode: 'spread' }>;
+
+/**
+ * Recursively merges two values, `a` and `b`.
+ * @param source - The second value to merge.
+ * @signature
+ *    R.mergeDeep(b)(a)
+ * @example
+ *    R.mergeDeep({ foo: 'baz', y: 2 })({ foo: 'bar', x: 1 }) // => { foo: 'bar', x: 1, y: 2 }
+ *    R.pipe(
+ *      { foo: 'bar', x: 1 },
+ *      R.mergeDeep({ foo: 'baz', y: 2 }),
+ *    ) // => { foo: 'bar', x: 1, y: 2 }
+ * @data_last
+ * @category Object
+ * @pipeable
+ */
+export function mergeDeep<
+  Target extends UnknownRecordOrArray,
+  Source extends UnknownRecordOrArray
+>(source: Source): (target: Target) => MergeDeep<Target, Source>;
+
+export function mergeDeep() {
+  return purry(_mergeDeep, arguments);
+}
+
+function _mergeDeep(a: UnknownRecordOrArray, b: UnknownRecordOrArray) {
+  const seenValues = new WeakSet(); // use to keep track of which objects have been seen.
+
+  function isArrayCyclic(arr: ReadonlyArray<unknown>) {
+    return arr.some(value => value && seenValues.has(value));
+  }
+
+  function merge(a: UnknownRecordOrArray, b: UnknownRecordOrArray) {
+    if (a === null || b === null) {
+      return a ?? b;
+    }
+
+    if (seenValues.has(a) || seenValues.has(b)) {
+      return a;
+    }
+    seenValues.add(a);
+    seenValues.add(b);
+
+    if (Array.isArray(a)) {
+      if (Array.isArray(b)) {
+        if (isArrayCyclic(a)) {
+          return b;
+        }
+        if (isArrayCyclic(b)) {
+          return a;
+        }
+        return a.concat(b);
+      }
+      return a;
+    }
+    if (Array.isArray(b)) {
+      return a;
+    }
+
+    // At this point the output is already merged, simply not deeply merged.
+    const output = { ...b, ...a };
+
+    // now just scan the output and look for values that should have been deep-merged
+    for (const k in b) {
+      const aValue = a[k];
+      const bValue = b[k];
+
+      if (
+        isRecordOrArray(aValue) &&
+        !seenValues.has(a[k] as object) &&
+        isRecordOrArray(bValue)
```

Comment:
We never recurse into arrays, this should only check for records as those are the only type that we actually perform a deep merge for. I'd even go as far as limit the whole logic of this method to just objects without any special handling of arrays (beyond spreading them if both values are arrays).
```suggestion
        isRecord(aValue) &&
        !seenValues.has(a[k] as object) &&
        isRecord(bValue)
```